s1kd-icncatalog/doc/s1kd-icncatalog.1
.\" Automatically generated by Pandoc 1.19.2.1
.\"
.TH "s1kd\-icncatalog" "1" "2018\-06\-01" "" "s1kd\-tools"
.hy
.SH NAME
.PP
s1kd\-icncatalog \- Manage the catalog used to resolve ICNs
.SH SYNOPSIS
.IP
.nf
\f[C]
s1kd\-icncatalog\ [options]\ [<object>...]
\f[]
.fi
.SH DESCRIPTION
.PP
The \f[I]s1kd\-icncatalog\f[] tool is used to manage a catalog of ICNs
for a project, and to resolve ICNs using this catalog.
Resolving an ICN means placing the actual filename of the ICN in to the
SYSTEM ID of the ENTITY declaration within CSDB objects.
.SH OPTIONS
.TP
.B \-a <ICN>
Add an ICN to the catalog.
Follow with the \-u and \-n options to specify the URI and notation to
use for this ICN.
The \-m option specifies a media group to add the ICN to.
.RS
.RE
.TP
.B \-c <catalog>
Specify the catalog file to manage or resolve against.
By default, the file "icncatalog.xml" in the current directory is used.
.RS
.RE
.TP
.B \-d <ICN>
Delete an ICN from the catalog.
The \-m option specifies a media group to delete the ICN from.
.RS
.RE
.TP
.B \-f
Overwrite the input CSDB objects when resolving ICNs, or overwrite the
catalog file when modifying it.
Otherwise, output is written to stdout.
.RS
.RE
.TP
.B \-m <media>
Resolve ICNs for this intended output media.
The catalog may contain alternative formats for the same ICN to be used
for different output media.
.RS
.RE
.TP
.B \-n <notation>
Specify the notation to reference when adding an ICN with the \-a
option.
.RS
.RE
.TP
.B \-t
Create a new empty catalog.
.RS
.RE
.TP
.B \-u <URI>
Specify the URI when adding an ICN with the \-a option.
.RS
.RE
.TP
.B \-x
Process input CSDB objects using the XInclude specification.
.RS
.RE
.TP
.B \-\-version
Show version information.
.RS
.RE
.SH EXAMPLES
.SS Resolving ICNs to filenames
.PP
A CSDB object may reference an ICN as follows:
.IP
.nf
\f[C]
<!NOTATION\ png\ SYSTEM\ "png">
<!ENTITY\ ICN\-TEST\-00001\-001\-01\ SYSTEM\ "ICN\-TEST\-00001\-001\-01.PNG"
NDATA\ png>
\f[]
.fi
.PP
The SYSTEM ID of this ENTITY indicates that the ICN file will be in the
same directory relative to the CSDB object.
However, the ICN files in this example are located in a separate folder
called \[aq]graphics\[aq].
Rather than manually updating every ENTITY declaration in every CSDB
object, a catalog file can be used to map ICNs to actual filenames:
.IP
.nf
\f[C]
<icnCatalog>
<icn\ infoEntityIdent="ICN\-TEST\-00001\-001\-01"
uri="graphics/ICN\-TEST\-00001\-001\-01.PNG"/>
</icnCatalog>
\f[]
.fi
.PP
Then, using this tool, the ICN can be resolved against the catalog:
.IP
.nf
\f[C]
$\ s1kd\-icncatalog\ \-c\ <catalog>\ <object>
\f[]
.fi
.PP
Producing the following output:
.IP
.nf
\f[C]
<!NOTATION\ png\ SYSTEM\ "png">
<!ENTITY\ ICN\-TEST\-00001\-001\-01\ SYSTEM
"graphics/ICN\-TEST\-00001\-001\-01.PNG"\ NDATA\ png>
\f[]
.fi
.SS Alternative ICN formats
.PP
A catalog can also be used to provide alternative file formats for an
ICN depending on the intended output media.
For example:
.IP
.nf
\f[C]
<icnCatalog>
<notation\ name="jpg"\ systemId="jpg"/>
<notation\ name="svg"\ systemId="svg"/>
<media\ name="pdf">
<icn\ infoEntityIdent="ICN\-TEST\-00001\-001\-01"
uri="ICN\-TEST\-00001\-001\-01.JPG"\ notation="jpg"/>
</media>
<media\ name="web">
<icn\ infoEntityIdent="ICN\-TEST\-00001\-001\-01"
uri="ICN\-TEST\-00001\-001\-01.SVG"\ notation="svg"/>
</media>
</icnCatalog>
\f[]
.fi
.PP
The \-m option allows for specifying which type of media to resolve for:
.IP
.nf
\f[C]
<!NOTATION\ png\ SYSTEM\ "png">
<!ENTITY\ ICN\-TEST\-00001\-001\-01\ SYSTEM\ "ICN\-TEST\-00001\-001\-01.PNG"
NDATA\ png>
\f[]
.fi
.IP
.nf
\f[C]
$\ s1kd\-icncatalog\ \-c\ <catalog>\ \-m\ pdf\ <object>
\f[]
.fi
.IP
.nf
\f[C]
<!NOTATION\ png\ SYSTEM\ "png">
<!NOTATION\ jpg\ SYSTEM\ "jpg">
<!ENTITY\ ICN\-TEST\-00001\-001\-01\ SYSTEM\ "ICN\-TEST\-00001\-001\-01.JPG"
NDATA\ jpg>
\f[]
.fi
.IP
.nf
\f[C]
$\ s1kd\-icncatalog\ \-c\ <catalog>\ \-m\ web\ <object>
\f[]
.fi
.IP
.nf
\f[C]
<!NOTATION\ png\ SYSTEM\ "png">
<!NOTATION\ svg\ SYSTEM\ "svg">
<!ENTITY\ ICN\-TEST\-00001\-001\-01\ SYSTEM\ "ICN\-TEST\-00001\-001\-01.SVG"
NDATA\ svg>
\f[]
.fi
.SH CATALOG SCHEMA
.PP
The following describes the schema of an ICN catalog file.
.SS Catalog
.PP
\f[I]Markup element:\f[] \f[C]<icnCatalog>\f[]
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
None
.PP
\f[I]Child elements:\f[]
.IP \[bu] 2
\f[C]<notation>\f[]
.IP \[bu] 2
\f[C]<media>\f[]
.IP \[bu] 2
\f[C]<icn>\f[]
.SS Notation
.PP
The element \f[C]<notation>\f[] represents a NOTATION declaration.
.PP
\f[I]Markup element:\f[] \f[C]<notation>\f[]
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
\f[C]name\f[], the NDATA name.
.IP \[bu] 2
\f[C]publicId\f[], the optional PUBLIC ID of the notation.
.IP \[bu] 2
\f[C]systemId\f[], the optional SYSTEM ID of the notation.
.PP
\f[I]Child elements:\f[]
.IP \[bu] 2
None
.SS Media
.PP
The element \f[C]<media>\f[] groups a set of alternative ICN formats for
a particular output media type.
.PP
\f[I]Markup element:\f[] \f[C]<media>\f[]
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
\f[C]name\f[], the identifier of the output media.
.PP
\f[I]Child elements:\f[]
.IP \[bu] 2
\f[C]<icn>\f[]
.SS ICN
.PP
The element \f[C]<icn>\f[] maps an ICN to a filename and optionally a
notation.
When this element occurs as a child of a \f[C]<media>\f[] element, it
will be used when that output media is specified with the \-m option.
When it occurs as a child of \f[C]<icnCatalog>\f[], it will be used if
no media is specified.
.PP
\f[I]Markup element:\f[] \f[C]<icn>\f[]
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
\f[C]infoEntityIdent\f[], the ICN
.IP \[bu] 2
\f[C]uri\f[], the filename the ICN will resolve to
.IP \[bu] 2
\f[C]notation\f[], a reference to a previously declared
\f[C]<notation>\f[] element.
.PP
\f[I]Child elements:\f[]
.IP \[bu] 2
None
.SS Example ICN catalog
.IP
.nf
\f[C]
<icnCatalog>
<notation\ name="jpg"\ systemId="jpg"/>
<notation\ name="png"\ systemId="png"/>
<notation\ name="svg"\ systemId="svg"/>
<media\ name="pdf">
<icn\ infoEntityIdent="ICN\-TEST\-00001\-001\-01"
uri="ICN\-TEST\-00001\-001\-01.JPG"\ notation="jpg"/>
</media>
<media\ name="web">
<icn\ infoEntityIdent="ICN\-TEST\-00001\-001\-01"
uri="ICN\-TEST\-00001\-001\-01.SVG"\ notation="svg"/>
</media>
<icn\ infoEntityIdent="ICN\-TEST\-00001\-001\-01"
uri="ICN\-TEST\-00001\-001\-01.PNG"\ notation="png"/>
</icnCatalog>
\f[]
.fi
.SH AUTHORS
khzae.net.


tools/s1kd-icncatalog/s1kd-icncatalog.c
#include <stdio.h>
#include <unistd.h>
#include <getopt.h>
#include <stdbool.h>
#include <string.h>
#include <libgen.h>
#include <libxml/tree.h>
#include <libxml/xinclude.h>
#include <libxml/valid.h>
#include <libxml/xpath.h>
#include "templates.h"

#define PROG_NAME "s1kd-icncatalog"
#define VERSION "1.0.0"

#define DEFAULT_CATALOG_NAME "icncatalog.xml"

/* Bug in libxml < 2.9.2 where parameter entities are resolved even when
 * XML_PARSE_NOENT is not specified.
 */
#if LIBXML_VERSION < 20902
#define PARSE_OPTS XML_PARSE_NONET
#else
#define PARSE_OPTS 0
#endif

/* Not exposed by the libxml API */
static void xmlFreeEntity(xmlEntityPtr entity)
{
    xmlDictPtr dict = NULL;

    if (entity == NULL)
        return;

    if (entity->doc != NULL)
        dict = entity->doc->dict;


    if ((entity->children) && (entity->owner == 1) &&
        (entity == (xmlEntityPtr) entity->children->parent))
        xmlFreeNodeList(entity->children);
    if (dict != NULL) {
        if ((entity->name != NULL) && (!xmlDictOwns(dict, entity->name)))
            xmlFree((char *) entity->name);
        if ((entity->ExternalID != NULL) &&
	    (!xmlDictOwns(dict, entity->ExternalID)))
            xmlFree((char *) entity->ExternalID);
        if ((entity->SystemID != NULL) &&
	    (!xmlDictOwns(dict, entity->SystemID)))
            xmlFree((char *) entity->SystemID);
        if ((entity->URI != NULL) && (!xmlDictOwns(dict, entity->URI)))
            xmlFree((char *) entity->URI);
        if ((entity->content != NULL)
            && (!xmlDictOwns(dict, entity->content)))
            xmlFree((char *) entity->content);
        if ((entity->orig != NULL) && (!xmlDictOwns(dict, entity->orig)))
            xmlFree((char *) entity->orig);
    } else {
        if (entity->name != NULL)
            xmlFree((char *) entity->name);
        if (entity->ExternalID != NULL)
            xmlFree((char *) entity->ExternalID);
        if (entity->SystemID != NULL)
            xmlFree((char *) entity->SystemID);
        if (entity->URI != NULL)
            xmlFree((char *) entity->URI);
        if (entity->content != NULL)
            xmlFree((char *) entity->content);
        if (entity->orig != NULL)
            xmlFree((char *) entity->orig);
    }
    xmlFree(entity);
}

/* Add a NOTATION */
void add_notation(xmlDocPtr doc, const xmlChar *name, const xmlChar *pubId, const xmlChar *sysId)
{
	xmlValidCtxtPtr valid;

	if (!doc->intSubset) {
		return;
	}

	if (!xmlHashLookup(doc->intSubset->notations, BAD_CAST name)) {
		valid = xmlNewValidCtxt();
		xmlAddNotationDecl(valid, doc->intSubset, name, pubId, sysId);
		xmlFreeValidCtxt(valid);
	}
}

/* Add a notation by its reference in the catalog file. */
void add_notation_ref(xmlDocPtr doc, xmlDocPtr icns, const xmlChar *notation)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;

	ctx = xmlXPathNewContext(icns);
	obj = xmlXPathEvalExpression(BAD_CAST "/icnCatalog/notation", ctx);

	if (!xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
		int i;
		bool found = false;

		for (i = 0; i < obj->nodesetval->nodeNr && !found; ++i) {
			xmlChar *name;

			name = xmlGetProp(obj->nodesetval->nodeTab[i], BAD_CAST "name");

			if (xmlStrcmp(name, notation) == 0) {
				xmlChar *pubId, *sysId;

				pubId = xmlGetProp(obj->nodesetval->nodeTab[i], BAD_CAST "publicId");
				sysId = xmlGetProp(obj->nodesetval->nodeTab[i], BAD_CAST "systemId");

				add_notation(doc, name, pubId, sysId);

				xmlFree(pubId);
				xmlFree(sysId);

				found = true;
			}

			xmlFree(name);
		}
	}

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);
}

/* Resolve the ICNs in a document against the ICN catalog. */
void resolve_icn(xmlDocPtr doc, xmlDocPtr icns, xmlChar *ident, xmlChar *uri, xmlChar *notation)
{
	xmlEntityPtr e;

	e = xmlGetDocEntity(doc, ident);

	if (e) {
		xmlChar *ndata;

		if (!notation) {
			ndata = xmlStrdup(e->content);
		}

		xmlUnlinkNode((xmlNodePtr) e);
		xmlFreeEntity(e);

		if (notation) {
			xmlAddDocEntity(doc, ident, XML_EXTERNAL_GENERAL_UNPARSED_ENTITY, NULL, uri, notation);
		} else {
			xmlAddDocEntity(doc, ident, XML_EXTERNAL_GENERAL_UNPARSED_ENTITY, NULL, uri, ndata);
		}

		if (notation) {
			add_notation_ref(doc, icns, notation);
		} else {
			xmlFree(ndata);
		}
	}
}

/* Resolve ICNs in a file against the ICN catalog. */
void resolve_icns_in_file(const char *fname, xmlDocPtr icns, bool overwrite, bool xinclude, const char *media)
{
	xmlDocPtr doc;
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;
	xmlChar xpath[256];

	doc = xmlReadFile(fname, NULL, PARSE_OPTS);

	if (xinclude) {
		xmlXIncludeProcess(doc);
	}

	ctx = xmlXPathNewContext(icns);

	if (media) {
		xmlStrPrintf(xpath, 256, "/icnCatalog/media[@name='%s']/icn", media);
	} else {
		xmlStrPrintf(xpath, 256, "/icnCatalog/icn");
	}

	obj = xmlXPathEvalExpression(xpath, ctx);

	if (!xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
		int i;

		for (i = 0; i < obj->nodesetval->nodeNr; ++i) {
			xmlChar *ident, *uri, *notation;

			ident    = xmlGetProp(obj->nodesetval->nodeTab[i], BAD_CAST "infoEntityIdent");
			uri      = xmlGetProp(obj->nodesetval->nodeTab[i], BAD_CAST "uri");
			notation = xmlGetProp(obj->nodesetval->nodeTab[i], BAD_CAST "notation");

			resolve_icn(doc, icns, ident, uri, notation);

			xmlFree(ident);
			xmlFree(uri);
			xmlFree(notation);
		}
	}

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);

	if (overwrite) {
		xmlSaveFile(fname, doc);
	} else {
		xmlSaveFile("-", doc);
	}

	xmlFreeDoc(doc);
}

xmlNodePtr first_xpath_node(xmlDocPtr doc, xmlNodePtr node, const xmlChar *expr)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;
	xmlNodePtr first;

	ctx = xmlXPathNewContext(doc ? doc : node->doc);
	ctx->node = node;

	obj = xmlXPathEvalExpression(expr, ctx);

	first = xmlXPathNodeSetIsEmpty(obj->nodesetval) ? NULL : obj->nodesetval->nodeTab[0];

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);

	return first;
}

void add_icns(xmlDocPtr icns, xmlNodePtr add, const char *media)
{
	xmlNodePtr root, cur;

	if (media) {
		xmlChar xpath[256];
		xmlStrPrintf(xpath, 256, "/icnCatalog/media[@name='%s']", media);
		root = first_xpath_node(icns, NULL, xpath);
	} else {
		root = xmlDocGetRootElement(icns);
	}


	for (cur = add->children; cur; cur = cur->next) {
		xmlAddChild(root, xmlCopyNode(cur, 1));
	}
}

void del_icns(xmlDocPtr icns, xmlNodePtr del, const char *media)
{
	xmlNodePtr cur;

	for (cur = del->children; cur; cur = cur->next) {
		xmlChar xpath[256];
		xmlChar *ident;
		xmlNodePtr icn;

		ident = xmlGetProp(cur, BAD_CAST "infoEntityIdent");
		if (media) {
			xmlStrPrintf(xpath, 256, "/icnCatalog/media[@name='%s']/icn[@infoEntityIdent='%s']", media, ident);
		} else {
			xmlStrPrintf(xpath, 256, "/icnCatalog/icn[@infoEntityIdent='%s']", ident);
		}
		xmlFree(ident);

		if ((icn = first_xpath_node(icns, NULL, xpath))) {
			xmlUnlinkNode(icn);
			xmlFreeNode(icn);
		}
	}
}

/* Help/usage message. */
void show_help(void)
{
	puts("Usage: " PROG_NAME " [options] [<object>...]");
	puts("");
	puts("Options:");
	puts("  -h -?  Show help/usage message.");
	puts("  -a <icn>       Add an ICN to the catalog.");
	puts("  -c <catalog>   Use <catalog> as the ICN catalog.");
	puts("  -d <icn>       Delete an ICN from the catalog.");
	puts("  -f             Overwrite input objects.");
	puts("  -m <media>     Specify intended output media.");
	puts("  -n <notation>  Set the notation of the new ICN.");
	puts("  -t             Create new ICN catalog.");
	puts("  -u <uri>       Set the URI of the new ICN.");
	puts("  -x             Process XInclude elements.");
	puts("  --version      Show version information.");
}

void show_version(void)
{
	printf("%s (s1kd-tools) %s\n", PROG_NAME, VERSION);
}

int main(int argc, char **argv)
{
	int i;
	bool overwrite = false;
	bool xinclude = false;
	char *icns_fname = NULL;
	bool createnew = false;
	char *media = NULL;
	xmlDocPtr icns;
	xmlNodePtr add, del, cur = NULL;

	const char *sopts = "a:c:d:fm:n:tu:xh?";
	struct option lopts[] = {
		{"version", no_argument, 0, 0},
		{0, 0, 0, 0}
	};
	int loptind = 0;

	add = xmlNewNode(NULL, BAD_CAST "add");
	del = xmlNewNode(NULL, BAD_CAST "del");

	while ((i = getopt_long(argc, argv, sopts, lopts, &loptind)) != -1) {
		switch (i) {
			case 0:
				if (strcmp(lopts[loptind].name, "version") == 0) {
					show_version();
					return 0;
				}
				break;
			case 'a':
				cur = xmlNewChild(add, NULL, BAD_CAST "icn", NULL);
				xmlSetProp(cur, BAD_CAST "infoEntityIdent", BAD_CAST optarg);
				break;
			case 'c':
				if (!icns_fname) {
					icns_fname = strdup(optarg);
				}
				break;
			case 'd':
				cur = xmlNewChild(del, NULL, BAD_CAST "icn", NULL);
				xmlSetProp(cur, BAD_CAST "infoEntityIdent", BAD_CAST optarg);
				break;
			case 'f':
				overwrite = true;
				break;
			case 'm':
				if (!media) {
					media = strdup(optarg);
				}
				break;
			case 'n':
				if (cur) {
					xmlSetProp(cur, BAD_CAST "notation", BAD_CAST optarg);
				}
				break;
			case 't':
				createnew = true;
				break;
			case 'u':
				if (cur) {
					xmlSetProp(cur, BAD_CAST "uri", BAD_CAST optarg);
				}
				break;
			case 'x':
				xinclude = true;
				break;
			case 'h':
			case '?':
				show_help();
				return 0;
		}
	}

	if (!icns_fname) {
		icns_fname = strdup(DEFAULT_CATALOG_NAME);
	}

	if (createnew || access(icns_fname, F_OK) == -1) {
		icns = xmlReadMemory((const char *) icncatalog_xml, icncatalog_xml_len, NULL, NULL, 0);
	} else {
		icns = xmlReadFile(icns_fname, NULL, PARSE_OPTS);
	}

	if (add->children || del->children) {
		if (add->children) {
			add_icns(icns, add, media);
		}
		if (del->children) {
			del_icns(icns, del, media);
		}
		if (overwrite) {
			xmlSaveFile(icns_fname, icns);
		} else {
			xmlSaveFile("-", icns);
		}
	} else if (optind < argc) {
		for (i = optind; i < argc; ++i) {
			resolve_icns_in_file(argv[i], icns, overwrite, xinclude, media);
		}
	} else if (createnew) {
		if (overwrite) {
			xmlSaveFile(icns_fname, icns);
		} else {
			xmlSaveFile("-", icns);
		}
	} else {
		resolve_icns_in_file("-", icns, false, xinclude, media);
	}

	free(icns_fname);
	xmlFreeNode(add);
	xmlFreeNode(del);
	xmlFreeDoc(icns);
	xmlCleanupParser();

	return 0;
}
